### ISSUE TITLE
Shuffling segment order while using ExternSprintDataset with returnn-rasr training

### ISSUE THREAD

**Issue by Marvin84** [rwth-i6/i6_core#110]
It is experimentally shown that setting the [`segment-order-shuffle`](https://p.i6.informatik.rwth-aachen.de/diffusion/RASR/browse/master/src/Bliss/CorpusDescription.cc$184) parameter of the Rasr `CorpusDescription` class to `true` can lead to WER improvement. Therefore, it is recommended to take this into account for all training setups relying on `ExternSprintDataset`. 

In this issue I would like to open a discussion on two important aspects:

1. What is the interaction/difference between segment order shuffling in Rasr and other data shuffling mechanisms during training in Returnn? 
2. What is the cleanest solution for setting this parameter to true in our setups, without breaking existing jobs and pipelines?

For the second point, @michelwi proposed to change the corpus setting in train`crp`. Another solution, as pointed out by @curufinwe, is to set this in the `extra_config` passed to the Job.

Let's open the discussion.

**Comment by albertz** [comment]
I'm not sure I understand. This is always enabled, or not?


Tina Raissi ***@***.***> schrieb am Di., 3. Aug. 2021, 17:49:

**Comment by Marvin84** [comment]
> This is always enabled, or not?

No, it is not. The Rasr default is false. On the setup side, we do not set it [here](https://github.com/rwth-i6/i6_core/blob/main/returnn/rasr_training.py#L268). I know that @mkitza for his hybrid setups set it via `extra_config`, and @ZhouW321 hardcode it in the string linked above in the `create_dataset_config` function.

**Comment by albertz** [comment]
> > This is always enabled, or not?
> 
> No, it is not. The Rasr default is false. On the setup side, we do not set it [here](https://github.com/rwth-i6/i6_core/blob/main/returnn/rasr_training.py#L268). I know that @mkitza for his hybrid setups set it via `extra_config`, and @ZhouW321 hardcode it in the string linked above in the `create_dataset_config` function.

Ok, I don't know any of that code. But in all my setups, you find that I set it to true.

But I'm also not exactly understanding your question/issue here. Or is this not really an issue but just for clarification / discussion?

Are you asking about the bucket ("laplace") shuffling in RETURNN? What's unclear about it? It's similar. (I don't remember the details of RASR shuffling anymore. Maybe the sorting was always ascending in a bucket? In RETURNN, it alternates between ascending and descending.)

On the generic question whether you should shuffle the segments (and a different order for every full epoch): Definitely yes. Otherwise it will overfit much more.

**Comment by Marvin84** [comment]
> But I'm also not exactly understanding your question/issue here

Thanks for the answers above. It is clear that shuffling is necessary, as you can see from the question itself. What I and others would like to make sure of is the interaction between this Rasr side shuffle and other shufflings in Returnn. I am not interested in how the shuffling is done in Returnn, but (1) where and when it is done, (2) whether the segment's order goes through another shuffle done by Returnn.

**Comment by albertz** [comment]
> I am not interested in how the shuffling is done in Returnn, but (1) where and when it is done, (2) whether the segment's order goes through another shuffle done by Returnn.

No, normally you cannot do both.

In RETURNN, the dataset implementation (`Dataset` sub class) does the shuffling. Every dataset could have its own logic for that. In case of `ExternSprintDataset`, there is no extra logic at all. So only RASR does it.

When `ExternSprintDataset` is used as part of another dataset (e.g. `MetaDataset`), some other dataset might control the sequence order. Then Sprint will be given a predefined sequence order file, and no RASR seq shuffling/sorting will be used.

**Comment by JackTemaki** [comment]
> But I'm also not exactly understanding your question/issue here.

I will try to break it down into clear questions and issues I am seeing (related to RETURNN now):


**Are there, and if yes, what are the ways to control the segment order of the RASR process via RETURNN?**
From what I see there is only the possibility to use the "filtering" option, but never to specify an order.

So for filtering there is:

```python
    if self.seq_tags_filter is not None:
      assert not self.predefined_seq_list_order
      import tempfile
      self.seq_list_file = tempfile.mktemp(prefix="returnn-sprint-predefined-seq-filter")
      with open(self.seq_list_file, "w") as f:
        for tag in self.seq_tags_filter:
          f.write(tag)
          f.write("\n")
        f.close()
      args += ["--*.corpus.segments.file=%s" % self.seq_list_file]
```
Which will be reached when specifying `seq_list_filter_file`, which is an "official" Dataset parameter.

There is also this code:
```python
if self.predefined_seq_list_order:
      import tempfile
      self.seq_list_file = tempfile.mktemp(prefix="returnn-sprint-predefined-seq-list")
      with open(self.seq_list_file, "w") as f:
        for tag in self.predefined_seq_list_order:
          f.write(tag)
          f.write("\n")
        f.close()
      args += [
        "--*.corpus.segment-order-shuffle=false",
        "--*.corpus.segments.file=%s" % self.seq_list_file,
        "--*.corpus.segment-order=%s" % self.seq_list_file]
```

Which would in theory allow to set a pre-defined order. But I found no way to ever activate this, so for me it seems this is dead code (at least if you work directly with ExternSprintDataset via configs, when using MetaDataset or RETURNN as a framework it is of course reachable, because you can e.g. call `init_seq_order()` by hand with `seq_list` defined).
 
Then `--*.corpus.segment-order-shuffle=false` was set to be sure it does not shuffle, so one could get the idea that it shuffles on default, but this is a rather minor issue.


** What is `python-segment-order` doing?**

This is especially about those three parameters that are automatically set:
```python
    args += [
      "--*.python-segment-order=true",
      "--*.python-segment-order-pymod-path=%s" % self._my_python_mod_path,
      "--*.python-segment-order-pymod-name=returnn.sprint.extern_interface",
      "--*.use-data-source=false",
      "--*.trainer=python-trainer",
      "--*.pymod-path=%s" % self._my_python_mod_path,
      "--*.pymod-name=returnn.sprint.extern_interface",
      "--*.pymod-config=%s" % config_str]
```

So the interface is needed in order for the ExternSprintDataset to work, but here I am not really sure what the `segment-order` arguments mean. This is also completely missing from the RASR documentation.

** ExternSpringDataset allows to set `seq_order` **

This is one of the problems that caused this whole confusion. People are setting `"seq_order": "random"` or `"batching": "random"`, and this does exactly nothing with no error from RETURNN. The ExternSprintDataset does not use `get_seq_order_for_epoch()`, so this is a dead parameter. So either this should crash, or at least support `random` with actually setting `--*.corpus.segment-order-shuffle=true` directly via the ExternSprintDataset.

We can discuss this in the RETURNN repo, if we want to do this with a `behavior_version` update.

**Comment by albertz** [comment]
I think this is resolved now, right?

Also, code is not dead if it is reachable. This was exactly the case I was explaining above.

> We can discuss this in the RETURNN repo, if we want to do this with a behavior_version update.

You mean whether setting `seq_order != "default"` on `ExternSprintDataset` should produce some exception? Yea, we can do that. (But this is for RETURNN, not here for i6_core.)

**Comment by JackTemaki** [comment]
> But this is for RETURNN, not here for i6_core

This is why I reopened this, we have to decide if we fix it via the recipes or via RETURNN (or both). And not only I should have an opinion about this and give comments about this. This issue was about discussing the solution. Until one is found, this issue is not resolved.

**Comment by albertz** [comment]
I don't exactly understand? Is this an issue here, or just a discussion thread, for clarification? (I asked this before but still not clear to me.) If this is an issue, what exactly is the issue? If this is for clarification, I think we clarified everything now, right?

**Comment by JackTemaki** [comment]
Please read the second point of the initial post again:

>  2. What is the cleanest solution for setting this parameter to true in our setups, without breaking existing jobs and pipelines?
> 
> 
> For the second point, @michelwi proposed to change the corpus setting in train`crp`. Another solution, as pointed out by @curufinwe, is to set this in the `extra_config` passed to the Job.
> 
> Let's open the discussion.

So this was about clarification AND a discussion thread to discuss the solution.

Currently the solution candidates are:
 - use `crp`
 - use `extra_config`
 - fix it in RETURNN and use `behavior_version`

**Comment by albertz** [comment]
> fix it in RETURNN and use behavior_version

But I don't understand. Fix what? Or change what? What is the actual proposal here for RETURNN? What is the behavior you want? Or what is the functionality you want?

**Comment by Marvin84** [comment]
>  Is this an issue here, or just a discussion thread, for clarification?

I thought it was clear from the labels that this is a question and discussion.

**Comment by JackTemaki** [comment]
> What is the actual proposal here for RETURNN?

If we go with the "RETURNN" solution, this would mean what I described above:
setting `"seq_order": "random"` for the `ExternSprintDataset` will set `--*.corpus.segment-order-shuffle=false` in `_build_sprint_args()`.

**Comment by albertz** [comment]
> > Is this an issue here, or just a discussion thread, for clarification?
> 
> I thought it was clear from the labels that this is a question and discussion.

But then this is resolved now, as it was answered?

Or when is this resolved? You want that everyone you assigned gives some opinion on this, and then it is resolved?

Or you want that we all agree on "one best solution" and then it is resolved? (I.e. never?)

**Comment by JackTemaki** [comment]
> Or when is this resolved?
It is resolved when `i6_core` supports shuffling for RETURNN-RASR training.

**Comment by albertz** [comment]
> > What is the actual proposal here for RETURNN?
> 
> If we go with the "RETURNN" solution, this would mean what I described above:
> setting `"seq_order": "random"` for the `ExternSprintDataset` will set `--*.corpus.segment-order-shuffle=true` in `_build_sprint_args()`.

Ah ok.
But I would not do that. I think this is un-intuitive. This can behave different to what you normally expect with `seq_order` (in subtle ways).
Also, you would not use that. You would want bucketing. E.g. what you have with "laplace". And the RASR logic is slightly different.
Only if you can guarantee that RASR is exactly behaving in the same way as RETURNN with other datasets for `seq_order`, I would do that. But this is easy to get wrong, so that is why I would not do it.

**Comment by Marvin84** [comment]
> But it does already, via some of the suggested options?

As pointed out in the initial question, the default setting does not shuffle. So we need to find an agreement what is the best way to set it without breaking old experiments.

> Currently the solution candidates are:
> 
> use crp
> use extra_config
> fix it in RETURNN and use behavior_version

I would stick to one of the first two options. Any argument pro or against any of them?

**Comment by JackTemaki** [comment]
> You would want bucketing.

No, this is not needed for chunking setups.

>  I think this is un-intuitive

But we should still fix in RETURNN that seq_order does not behave as expected (it should crash when setting it).

**Comment by JackTemaki** [comment]
> I would stick to one of the first two options. Any argument pro or against any of them?

I first thought that it would be fine to add this directly with `set_corpus` to the train crp, but I am not sure if this has any consequences when doing alignments...

With `extra_rasr_config` we could maybe set it in `train_nn`? For me it is really difficult to say...

**Comment by michelwi** [comment]
>Currently the solution candidates are:
>
>use crp
>use extra_config


>Any argument pro or against any of them?

Even though having suggested this, one argument against crp would be that it is in fact not a common RASR parameter, as it is only used in NN training and not in e.g. feature extraction, GMM training, lattice generation,

One argument against `extra_rasr_config` would be that it is easy to miss when building setups and awkward to have extra opts for a parameter that effectively should be set by default.

One argument in favor of crp is, that we have a special overlay for NN training anyway, so we can set it only there and leave the parent crp without this option. (this is currently my favourite)


I would like to add an additional option to the discussion: The `ReturnnRasrTrainingJob` can take an additional argument and add the shuffling while creating the dataset.
It could even read out the `batching="random"` option from the returnn config and act accordingly. (But this breaks existing setups, so I argue against it)

**Comment by JackTemaki** [comment]
> One argument against `extra_rasr_config` would be that it is easy to miss when building setups and awkward to have extra opts for a parameter that effectively should be set by default.

But this is maybe okay if we set this in the `train_nn` per default or not? Then nobody will miss this, but has to put effort in to not use it.

**Comment by michelwi** [comment]
>But this is maybe okay if we set this in the train_nn per default or not?

Sorry, what is "the `train_nn`" ? 
Do you mean the dict that contains the default args for the ReturnRasrTrainingJob, or e.g. the function in `meta.system` that initiates the NN training?

But then still people that are not using this function might miss it.

**Comment by Marvin84** [comment]
> But then still people that are not using this function might miss it.
 
I agree with Willi. This should be the responsibility of `ReturnnRasrTraining` class, together with all other training options. There might be cases where one does not rely on `meta.system` for training.

**Comment by JackTemaki** [comment]
> There might be cases where one does not rely on meta.system for training

Okay, if we assume this then really the Job parameter option is the only that makes sense. So then you will do an PR and we discuss the name there?

**Comment by michelwi** [comment]
>This should be the responsibility of ReturnnRasrTraining class [...]
 There might be cases where one does not rely on meta.system for training.

But even without `meta.system` a crp object is required so we could read it off of this object. This would not necessitate any change in our `i6_core` code only the training recipe has to be changed slightly. 
People who are not using `meta.system` still have to build their crp object somewhere and then have to take care of adding the shuffling on their own.

Besides, we could add a warning in the job/log when no shuffling is used.

**Comment by JackTemaki** [comment]
> But even without meta.system a crp object is required so we could read it off of this object

But then again people would (and will) miss it. If it is a parameter of the training job it can be set to default True, and people will not miss it.

**Comment by albertz** [comment]
We maybe could also introduce a behavior_version to i6_core, which would
allow for behavior changes, eg changing the default to shuffle sequences in
training.


michelwi ***@***.***> schrieb am Mi., 4. Aug. 2021, 12:12:

**Comment by JackTemaki** [comment]
I think the important point here is that shuffling needs to be opt-out, not opt-in. Or am I seeing this wrong?

**Comment by albertz** [comment]
Depending on the situation, the type of shuffling is also not obvious, eg
bucket size.

Maybe there just should not be any default, and always require the user to
set this? (Via new behavior version in i6 core.)


Nick Rossenbach ***@***.***> schrieb am Mi., 4. Aug. 2021,
12:15:

**Comment by Marvin84** [comment]
> if we assume this then really the Job parameter option is the only that makes sense.

Considering Willi's argument against `extra_rasr_config` and Nick's argument against `set_corpus`, I suggest that in the `ReturnnRasrTrainingJob` we read the value from `crp`. 

>I think the important point here is that shuffling needs to be opt-out, not opt-in.

Yes, but also not breaking old hashes. I think we can use Willi's suggestion about the warning and then take care that in the `i6_experiments` our pipelines using `ExternSprintDataset` set the parameter for the overlayed `crp`

**Comment by michelwi** [comment]
> I think the important point here is that shuffling needs to be opt-out, not opt-in. Or am I seeing this wrong?

Well, this is just a training option that we think will always be beneficial. But this is the case for many training options. Will you start enforcing the use of `chunking` in the train job? what about SpecAugment, etc.?

I agree that it is important to use, just do we need an explicit option for that if it can be set otherwise.


>  the type of shuffling is also not obvious, eg bucket size.

Yes, e.g. for sequence training we want it locally sorted.

**Comment by Marvin84** [comment]
> Depending on the situation, the type of shuffling is also not obvious, eg
> bucket size.

Yes. However, if as you said the segment shuffling, in this case, is done only by Rasr then this needs to be implemented in Rasr. I wonder if currently there is the possibility of having different options, @michelwi , do we?
I think this needs another discussion thread, though.

**Comment by JackTemaki** [comment]
> Yes, but also not breaking old hashes

Changing the default would make a default hash change, but not break the hash for shuffling=False.

> I agree that it is important to use, just do we need an explicit option for that if it can be set otherwise.
> I suggest that in the `ReturnnRasrTrainingJob` we read the value from `crp`.

But then we need to do exactly: nothing. I thought we wanted to make this more clear to the user....

**Comment by michelwi** [comment]
>  segment shuffling [...] is done only by Rasr 
I wonder if currently there is the possibility of having different options

Yes. e.g.
```
for c in ['returnn_train', 'returnn_dev']:
    system.crp[c].corpus_config.segment_order_sort_by_time_length = True
    system.crp[c].corpus_config.segment_order_sort_by_time_length_chunk_size = 1000
````

Is what I use for sequence training.


> I thought we wanted to make this more clear to the user.

This is the reason of this discussion. Doing nothing is one viable option. We might also include a warning for this easy to miss option.

**Comment by Marvin84** [comment]
> Changing the default would make a default hash change, but not break the hash for shuffling=False.

but again, the idea was that if a user today downloads the setups and the recipes, can run with no problem old experiments. Given this constraint, how can we warn/help the user to understand that the old recipe is problematic in this regard.

**Comment by curufinwe** [comment]
1. If you do not want to break/change old setups, you cannot change the default. You can exactly pick one of those two. The policy for the recipes is to not break/modify existing setups, so we must not change the default behavior
2. Regarding where to put it: to me it feels common enough to add it as a separate argument to RasrReturnnTrainingJob for more user convenience. Then the job can modify the crp.corpus_config (during config creation) to add this option.

**Comment by michelwi** [comment]
>  add it as a separate argument to RasrReturnnTrainingJob

this would necessitate a `__sis_hash_exclude__` right?
I'd prefer the crp solution then, if only for this.

**Comment by JackTemaki** [comment]
> this would necessitate a `__sis_hash_exclude__` right?

No, because the Job as a custom function which does its own logic.

**Comment by michelwi** [comment]
>No, because ...

ah, ok, then I am fine with either solution with only a slight preference to crp.

**Comment by curufinwe** [comment]
Closed due to inactivity.
